From 16756f02ab1cee5655b1d0cb6b2961c48118c744 Mon Sep 17 00:00:00 2001
From: Ravi Minnikanti <rminnikanti@marvell.com>
Date: Thu, 16 Apr 2026 20:25:12 +0530
Subject: [PATCH] lib, build: support gperftools without tc_mallinfo2

gperftools versions before 2.16 expose tc_mallinfo() but not tc_mallinfo2(). Detect the newer API at configure time and retain the compatible fallback used by SONiC builds.

Original-SONiC-Patch: 0112-SONiC-ONLY-lib-build-add-configure-check-for-tc_mallinfo2-and-fallback.patch

Signed-off-by: Ravi Minnikanti <rminnikanti@marvell.com>
---
 configure.ac | 5 +++++
 lib/vty.c    | 8 ++++++++
 2 files changed, 13 insertions(+)

diff --git a/configure.ac b/configure.ac
index 8778c24f2d..5ecda26f9d 100644
--- a/configure.ac
+++ b/configure.ac
@@ -920,6 +920,11 @@ AC_DEFINE([HAVE_TCMALLOC], [1], [Enable tcmalloc])
   *) AC_MSG_ERROR([bad value ${enableval} for --enable-gperf-tcmalloc]) ;;
 esac],[gperf_tcmalloc_enabled=false])
 
+if test "$gperf_tcmalloc_enabled" = "true"; then
+  AC_CHECK_FUNC([tc_mallinfo2],
+    [AC_DEFINE([HAVE_TC_MALLINFO2], [1], [tc_mallinfo2 from gperftools])])
+fi
+
 #if openssl, else use the internal
 AS_IF([test "$with_crypto" = "openssl"], [
 AC_CHECK_LIB([crypto], [EVP_DigestInit], [LIBS="$LIBS -lcrypto"], [], [])
diff --git a/lib/vty.c b/lib/vty.c
index 07b2251da2..27ddb2df10 100644
--- a/lib/vty.c
+++ b/lib/vty.c
@@ -3334,7 +3334,11 @@ DEFUN_NOSH (show_tcmalloc_stats,
 	    "tcmalloc library info\n"
 	    "Show tcmalloc stats\n")
 {
+#ifdef HAVE_TC_MALLINFO2
 	struct mallinfo2 minfo;
+#else
+	struct mallinfo minfo;
+#endif
 	char buf[30];
 	size_t sval;
 	int major = 0, minor = 0;
@@ -3344,7 +3348,11 @@ DEFUN_NOSH (show_tcmalloc_stats,
 
 	vty_out(vty, "Tcmalloc version '%s' (%d.%d)\n", str, major, minor);
 
+#ifdef HAVE_TC_MALLINFO2
 	minfo = tc_mallinfo2();
+#else
+	minfo = tc_mallinfo();
+#endif
 
 	vty_out(vty, "Tcmalloc mallinfo statistics:\n");
 	vty_out(vty, "  Total heap allocated:  %s\n",
